docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,323 @@
+**************
+Type Narrowing
+**************
+
+Python programs often contain symbols that take on multiple types within a
+single given scope and that are distinguished by a conditional check at
+runtime. For example, here the variable *name* can be either a ``str`` or
+``None``, and the ``if name is not None`` narrows it down to just ``str``::
+
+    def maybe_greet(name: str | None) -> None:
+        if name is not None:
+            print("Hello, " + name)
+
+This technique is called *type narrowing*.
+To avoid false positives on such code, type checkers understand
+various kinds of conditional checks that are used to narrow types in Python code.
+The exact set of type narrowing constructs that a type checker understands
+is not specified and varies across type checkers. Commonly understood
+patterns include:
+
+* ``if x is not None``
+* ``if x``
+* ``if isinstance(x, SomeType)``
+* ``if callable(x)``
+
+In addition to narrowing local variables, type checkers usually also support
+narrowing instance attributes and sequence members, such as
+``if x.some_attribute is not None`` or ``if x[0] is not None``, though the exact
```

Comment:
Is it worth saying that such narrowing is potentially unsound but still useful in practice?

Reply:
As the document explains later, virtually all narrowing is potentially unsound.

Reply:
It might still be useful to explain where the unsoundness comes from. E.g. I think something like `if x is not None` can only be unsound in the presence of dynamic features, such as `globals()["x"] = ...`.

Reply:
Thanks, I somehow missed it during my initial review. The section looks good.